GH-1155: Added support to handle the connection issues gracefully.

kittaakos opened the pull request:
Signed-off-by: Akos Kitta <kittaakos@gmail.com>

Closes #1155.

Please try this out. The connection can be tuned here:
<img width="275" alt="screen shot 2018-02-01 at 16 21 41" src="https://user-images.githubusercontent.com/1405703/35686215-0d4564e6-076c-11e8-938c-d7363b937b35.png">

I look forward to the feedbacks. Thanks for the help!

hexa00 commented:
@kittaakos btw I tested a bit (without your PR) Chrome's throttling feature and it doesn't apply to websockets.

I think it doesn't matter really for your PR but something to keep in mind if we want to test Theia in general we can't use that...

AlexTugarev commented on packages/core/src/common/connection-status.ts:
```diff
@@ -0,0 +1,42 @@
+ */
+
+/**
+ * Clients might contribute extensions and react on connection status changes.
```
Why not providing a service to let clients listen for such `Event`s?

AlexTugarev replied:
I see.

AlexTugarev commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,221 @@
+    }
+
+    onStart() {
+        this.timer = setInterval(() => {
```
just bikeshedding, but if you use `window.setInterval` here, then `timer` can be `number`.

akosyakov replied:
Does `setInterval` waits till the previous request is finished?

AlexTugarev replied:
don't think so. I could happen, that the pending request is not finished due to a very bad connection, but the next try is made.

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,221 @@
+    onStatusChange(event: ConnectionStatusChangeEvent) {
+        this.statusBar.removeElement('connection-status');
+        const text = `$(${this.getStatusIcon(event.health)})`;
+        const tooltip = event.health ? `Connection health: ${event.health}%` : 'Not connected';
```
`Online` and `Offline`? I think it is more understandable.

kittaakos replied:
I think you are absolutely right. I have changed it.

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,221 @@
+    }
+
+    private getStatusIcon(health: number | undefined) {
+        if (health === undefined || health === 0) {
```
maybe just green, yellow, read light?

kittaakos replied:
It is not possible :( We can have text and something font-awesome.

akosyakov replied:
I see :(

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,221 @@
+                    this.contributionProvider.getContributions().forEach(contribution => contribution.onStatusChange(this.connectionState));
+                }
+            };
+            xhr.onerror = () => {
```
i see that `xhr` has other events, like `ontimeout`, should not we handle them as well?

kittaakos replied:
OK.

akosyakov commented on packages/core/src/common/connection-status.ts:
```diff
@@ -0,0 +1,42 @@
+ * The connection-status states.
+ */
+export enum ConnectionState {
+    CONNECTED,
```
`ONLINE`, `OFFLINE`?

kittaakos replied:
Changed.

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,221 @@
+            this.state = event.state;
+            switch (event.state) {
+                case ConnectionState.CONNECTION_LOST: {
+                    const message = 'Application connection to the Theia backend is lost. Performing retries...';
```
`Theia appears to be offline. Attempting to reconnect...`?

akosyakov replied:
not sure about using Theia, since a concrete app can have another name, maybe `Your device` or `Your computer`

kittaakos replied:
You're right, I got rid of Theia.

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,221 @@
+                    break;
+                }
+                case ConnectionState.CONNECTED: {
+                    const message = 'Application connection to the Theia backend was successfully re-established.';
```
`Theia is online.`?

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,221 @@
+                this.contributionProvider.getContributions().forEach(contribution => contribution.onStatusChange(this.connectionState));
+            };
+            xhr.open('GET', `${window.location.href}alive`);
+            try { xhr.send(); } catch { /* NOOP */ }
```
why noop?

kittaakos replied:
Handled it.

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,221 @@
+            hasConnection = newHistory.slice(-this.props.threshold).some(s => s);
+        }
+        // Ideally, we do not switch back to online if we see any `true` items but, let's say, after three consecutive `true`s.
+        return new ConnectionStateMachine(this.props, hasConnection ? ConnectionState.CONNECTED : ConnectionState.CONNECTION_LOST, newHistory);
```
Do we need a new object each 2 sec?

kittaakos replied:
I do not understand your concern with this; we would create a new event instance anyway.

akosyakov replied:
ok

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,221 @@
+    public readonly value: void;
+
+    constructor() {
+        super({ title: 'Not connected' });
```
title: `Theia appears to be offline.`

AlexTugarev replied:
one could read this as "Theia server is offline", right? e.g. browsers say something like "there is no connection", if host is not reachable.

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,221 @@
+    constructor() {
+        super({ title: 'Not connected' });
+        this.closeCrossNode.remove();
+        const textNode = document.createTextNode('Application connection to the Theia backend is lost. Performing retries...');
```
`Attempting to reconnect...`?

kittaakos replied:
OK.

kittaakos commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,221 @@
+
+}
+
+class ConnectionStatusDialog extends AbstractDialog<void> {
```
I cannot see any benefit. Do you have at least one?

akosyakov replied:
Someone want to rebind `ApplicationConnectionStatusContribution` and override `getOrCreateDialog`?

kittaakos replied:
OK, make sense. 👍

kittaakos commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,221 @@
+                    this.messageService.info(message);
+                    if (this.dialog !== undefined) {
+                        // tslint:disable-next-line:no-any
+                        (<any>this.dialog).accept();
```
I did not want to dispose it, just reuse it.

akosyakov replied:
ok, just if we remove `accept` there won't be a compiler error, I think opening a new dialog is fine.

kittaakos replied:
OK.

kittaakos commented on packages/core/src/node/backend-connection-status.ts:
```diff
@@ -0,0 +1,36 @@
+    onStart(server: http.Server): void {
+        const __this = this;
+        server.once('listening', () => {
+            __this.app!.get('/alive', (request, response) => response.send('OK'));
```
I am not sure whether we have the `server` by the time we invoke `configure`. Let me check.

akosyakov replied:
but you don't need `server`

kittaakos replied:
Yes and no. I cannot log the address if I do not have the server.

akosyakov replied:
I meant only this line, you don't log an address there. `configure` is called to install handler, `onStart` is called when a server is already started with preinstalled handlers

kittaakos replied:
I have simplified it.

akosyakov commented on packages/core/src/node/backend-connection-status.ts:
```diff
@@ -0,0 +1,36 @@
+            __this.app!.get('/alive', (request, response) => response.send('OK'));
+            const { address, port } = server.address();
+            const host = `http://${address}:${port}`;
+            this.logger.info(`Started connection status endpoint at ${host} GET /alive.`);
```
that's not necessarily true, an app can be deployed on some path of host

kittaakos commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,221 @@
+                this.connectionState = this.connectionState.next(false);
+                this.contributionProvider.getContributions().forEach(contribution => contribution.onStatusChange(this.connectionState));
+            };
+            xhr.open('GET', `${window.location.href}alive`);
```
Thanks, I was not aware of this.

akosyakov commented on packages/core/src/node/backend-connection-status.ts:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    onStart(): void {
+        this.app!.get('/alive', (request, response) => response.send());
```
I meant `configure` is callback which should be used to install handlers, like: 
```ts
configure(app: express.Application): void {
    app.get('/alive', (request, response) => response.send());
}
```

kittaakos replied:
I see, of course. Thank you!

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,278 @@
+    }
+
+    onStart() {
+        this.timer = window.setInterval(() => {
```
Could it happen that there are several alive requests in parallel? Do we want it?

kittaakos replied:
> Could it happen that there are several alive requests in parallel?

Yes, it can. 

> Do we want it?

I do not think this will cause a problem but for the sake of simplicity, I change it.

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,278 @@
+    }
+
+    onStart() {
+        this.timer = window.setInterval(() => {
```
minor, we can do when someone request:
It would be nice to have protected method instead of lambdas to allow changing the default strategy, like `checkAlive()`, and provide reusable methods to compose a new strategy, like `updateStatus(success: boolean)`, `fireStatusChange()`.

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,278 @@
+ */
+export const ConnectionStatusService = Symbol('ConnectionStatusService');
+export interface ConnectionStatusService {
+    readonly onStatusChange: Event<ConnectionStatusChangeEvent>;
```
I think getter is missing for a current state in the case if someone wants to subscribe after the first event was fired.

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,278 @@
+ * Just in case we need to support a bit more sophisticated state transitions than having connection or not.
+ * For instance, `pending`, `reconnecting`.
+ */
+class ConnectionStateMachine implements ConnectionStatusChangeEvent {
```
minor: I think it should be re-exported if someone wants to modify defaults by rebinding `FrontendConnectionStatusService `

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,278 @@
+        return new ConnectionStateMachine(this.props, online ? ConnectionState.ONLINE : ConnectionState.OFFLINE, newHistory);
+    }
+
+    private updateHistory(success: boolean) {
```
minor: i would prefer `protected` as default to let subclasses access or override it

hexa00 commented:
Seems to work ok... a question for maybe further optimization I wonder if it would be possible to know that the user is idle.. i.e. No keyboard or mouse input and in that case suspend the alive check ?

That way 2-3 tabs staying up all night would not hit the server...

kittaakos commented:
@hexa00, this is a good idea. I try to implement it within this PR.

kittaakos commented:
@hexa00, I have pushed the changes. Currently, the frontend idle service activates after 15 seconds. Please reduce it if you would like to try that out.

hexa00 commented:
@akosyakov nice thx! 15 secs seems ok i"m trying it out

hexa00 commented:
Tested it works fine, I like the bed icon too :)

kittaakos commented:
@akosyakov, @hexa00, please take another look at the changes.

kittaakos commented:
> could we open issues about making the timeouts configurable once the preference-api is merged in core

Be my guest.

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,356 @@
+
+    onStop() {
+        if (this.timer !== undefined) {
+            window.clearInterval(this.timer);
```
`window.clearTimeout`

kittaakos replied:
Thanks!

akosyakov commented on packages/core/src/browser/frontend-idle-status.ts:
```diff
@@ -0,0 +1,136 @@
+
+    protected clearTimer() {
+        if (this.timer !== undefined) {
+            window.clearTimeout(this.timer);
```
`window.clearInterval`

kittaakos replied:
No, that is correct. The counterpart is the broken one. Thanks!

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,356 @@
+@injectable()
+export class ConnectionStatusStatusBarContribution extends DefaultFrontendApplicationContribution {
+
+    private idle: boolean;
```
Why don't get it from `idleService.idle`?

kittaakos replied:
I wanted to save the state.

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,356 @@
+        });
+    }
+
+    protected onStatusChange(event: ConnectionStatus) {
```
I think this method should be called both on `connectionStatusService.onStatusChange` and `idleService.onStatusChange`, remove `connection-status`, retrieve current values using getters and then render based on them

kittaakos replied:
I think the idle status should be merged into the connection status, new enumeration type (IDLE) has to be introduced.

akosyakov replied:
fine with me, I was mostly concerned about keeping the logic of rendering `connection-status` in one place

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,356 @@
+    }
+
+    protected onStatusChange(event: ConnectionStatus): void {
+        if (this.state !== event.state) {
```
It can happen that on `connectionStatusService.onStatusChange` is called with the same value several times? I think it should be called only when a value changed that clients don't do prev state caching and comparison like here.

kittaakos replied:
As you wish.

kittaakos replied:
This has to be there. It will not change.

kittaakos replied:
Yeah, otherwise, I cannot adjust the log message. I cannot make difference if the state was idle or offline previously after receiving the online state. Furthermore, the connection health which is a percentage between 0 and 100 can change while the state is still online. I hope it explains.

kittaakos replied:
Or I can provide another hook so that we would have both the `onConnectionStatusChange` and an `onConnectionStatusStateChange`.

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,356 @@
+        if (this.timer !== undefined) {
+            window.clearInterval(this.timer);
+            this.timer = undefined;
+            this.logger.info('Stopped backend connection status check.');
```
`debug`? i don't think it is useful for an end user

kittaakos replied:
Leftover.

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,356 @@
+    onStart() {
+        this.schedule(this.checkAlive.bind(this));
+        this.fireStatusChange(this.connectionState);
+        this.logger.info('Started backend connection status check.');
```
`debug`? i don't think it is useful for an end user

kittaakos replied:
Leftover.

marcdumais-work commented on packages/core/src/browser/frontend-idle-status.ts:
```diff
@@ -0,0 +1,136 @@
+     * The default idle options. The application goes into idle state after 15 seconds.
+     */
+    static DEFAULT: IdleStatusOptions = {
+        idleAfter: 15000
```
I like the idea of stopping the connection status checks when the frontend is idle (e.g. cloud instance that might be used sporadically). However I find 15 sec to be a short delay to be considered idle. If it were configurable, I would not mind. I would suggest 1 minute or more instead.

In my tests, I can lose the connection, and the idle mode kicks-in before it's detected that the connection was lost (assuming I do not move the mouse in the browser window or type in it).

marcdumais-work replied:
I also noticed that with the current idle timer, idle mode can kick-in while the user is patiently waiting for the connection to be detected as being up again. In that case, unless I generate bogus activity to prevent Theia from going into idle mode, it will never be detected that the connection has recovered.

e.g. in this video the connection is available when I dismiss the notification: 

![peek 2018-02-06 07-11](https://user-images.githubusercontent.com/25749063/35859139-f0a34688-0b0c-11e8-9015-b800925ab84a.gif)

kittaakos replied:
Thanks! I will split up the PR, open a follow-up and leave the idle feature out.

marcdumais-work replied:
Ok, if you think there need to be more discussions about the length of the idle timer. Else, with a longer timer, I think it would work fine.

kittaakos replied:
I think it is a useful and cool feature we have to support; I just need to wrap up other tasks too.

hexa00 replied:
@kittaakos please keep it in and just increase the idle timeout to the greater than the connection timeout , @marcdumais-work also agrees with this

kittaakos replied:
Sure, if you guys like this feature that much, we keep it as part of this PR :)

marcdumais-work replied:
Thanks @kittaakos - it is a nice feature.

kittaakos replied:
The credit goes to @hexa00.

kittaakos replied:
I have (force) pushed the changes regarding the idle service. Could you please verify the UX again? Thank you!

marcdumais-work replied:
ok, I will try it right away

marcdumais-work replied:
Works well, thanks @kittaakos

marcdumais-work commented:
Why have both a dialog and a user notification when we detect that the connection is lost? 
![image](https://user-images.githubusercontent.com/25749063/35858541-d0a16010-0b0a-11e8-917a-d21fab90c0cb.png)

At first look, I think the dialog would be enough. If the connection comes back, the dialog is automatically dismissed and the user doesn't have to wait to use Theia. With the notification, the user might not understand that it need to be dismissed first, before the connection can be tested again. If we had https://github.com/theia-ide/theia/issues/1081 , it would not be so bad.

marcdumais-work commented:
BTW, about my comment above (https://github.com/theia-ide/theia/pull/1186#issuecomment-363402524), I am ok to discuss this in an issue, after the PR is merged.

kittaakos commented:
> BTW, about my comment above (#1186 (comment)), I am ok to discuss this in an issue, after the PR is merged.

I think you are right, I will remove the redundant `messageService` notification. We will have the dialog.

akosyakov commented on packages/core/src/common/idle.ts:
```diff
@@ -0,0 +1,44 @@
+ * Service providing information whether it is in idle state or not.
+ */
+export const IdleService = Symbol('IdleService');
+export interface IdleService {
```
Do we need it in common? Could it be used in node?

akosyakov commented on packages/core/src/common/connection-status.ts:
```diff
@@ -0,0 +1,241 @@
+ * Service for listening on backend connection changes.
+ */
+export const ConnectionStatusService = Symbol('ConnectionStatusService');
+export interface ConnectionStatusService {
```
Do we need it in common? Could it be used in node?

kittaakos replied:
It should be easier to test without the browser.

akosyakov replied:
It will be testable without browser if it belongs to browser module as well?

Something should belong to common if it is supposed to be used in node and browser. If it is supposed to be used only in browser and electon-broswer, it should go to browser.

kittaakos replied:
> Something should belong to common if it is supposed to be used in node and browser

Exactly. It is an abstract service, you can use it from node too. If you implement the abstract methods.

akosyakov replied:
Do you think it is going to happen?

kittaakos replied:
Not at this point, but unless one can clearly state it will not happen, it can happen. It was easier for me to write tests without the browser. Did not check them in yet. Bear with me. I need to adjust two more things on the API.

akosyakov replied:
ok, just we do oppositive in other places, like planning moving commands, keybindings and menus to browser to expose apis from modules to which they belong

akosyakov commented on packages/core/src/browser/frontend-connection-status.ts:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    protected onStatusChange(status: ConnectionStatus): void {
+        if (this.state !== status.state) {
```
ok, have not thought about health rate

kittaakos replied:
I will change this to have to individual hooks. One for the state and one for the status changes.

akosyakov commented on packages/core/src/common/idle.ts:
```diff
@@ -0,0 +1,44 @@
+ */
+
+import { injectable } from 'inversify';
+import { Event } from '../common/event';
```
the file should be named after the main exported type

kittaakos replied:
Why? Since when do we have this policy?

akosyakov replied:
i think it always was, please check the coding guideline, it should be somewhere there

akosyakov replied:
https://github.com/theia-ide/theia/wiki/Coding-Guidelines#names:
> Name files after the main Type it exports

kittaakos replied:
Ohh, come on... I will change if you feel better:

```
 - promise-util.ts
 - cli.ts
 - extension-protocol.ts
 - ...
```

akosyakov replied:
we don't always pay attention, it is true, also utils and protocol are special since there is no a main concept

akosyakov commented on packages/core/src/browser/frontend-idle.ts:
```diff
@@ -0,0 +1,102 @@
+import { IdleService, IdleStatusOptions } from '../common/idle';
+
+@injectable()
+export class FrontendIdleService implements IdleService, FrontendApplicationContribution {
```
the file should be named after the main exported type

akosyakov commented on packages/core/package.json:
```diff
@@ -62,6 +62,7 @@
     "build": "theiaext build",
     "watch": "theiaext watch",
     "test": "theiaext test",
+    "test:watch": "theiaext test:watch",
```
@kittaakos I find it's strange:
- put a spec test file not next to a tested file
- to declare frontend APIs in common instead of the browser. Although I think that a separation of APIs and implementation is good, but it can be done by a new folder with 2 files under the browser folder:
  - one file with APIs reexported by this folder and by the browser folder
  - another file with the implementation which is used within the browser folder

I think that both APIs and tests should be put in browser, jsdom should be used to mock `window.setTimeout` instead of introducing an abstract superclass then everything can be encapsulated within the browser folder, not to be spread across the core extension.

Other opinions?

epatpol replied:
Yes I agree, as we've already started using `jsdom` in other places where frontend tests would fail because of the node environment, it would make sense to have the `.spec.ts` next to the service being tested.

akosyakov replied:
@svenefftinge can you review when you have time?

We need to adjust documentation if we change rules how tests and code should be located.

svenefftinge commented:
I'm afraid we conflated two things that don't belong together: idleness and connectivity.

I think the 'idle' feature doesn't belong here and should be removed. Why is it good to not ping the server when the browser is invisible? I want to see the offline status immediately when I come back. OTOH an unused idle backend should probably timeout. But it is a different topic/feature. Such idleness has to be communicated to the workspace server so it can stop the container. To detect that sort of idleness a simple lack of user interaction would not be enough as we would need to check running terminals and so on, too. Implementing a strategy for idleness can be very specific to how the workspace is hosted, too.

Also, I don't like to see this information in the status bar as I don't think it is useful to the user. She probably knows best that she hasn't been active on that window. But also it doesn't matter, does it?.

I think this PR would be smaller and simpler if we left the 'idleness' out and simply focus on the offline/online status.

kittaakos commented:
I have applied the following changes:
 - removed the idle service
 - moved the connection status service from common to the browser folder
 - got rid of the node tests configured tests to use jsdom

Please check the changes again. Thanks!

hexa00 commented:
@svenefftinge just to answer some questions:

> Why is it good to not ping the server when the browser is invisible? 
It just avoids unnecessary processing on the server side. And on the client side too.

> I want to see the offline status immediately when I come back.
That actually might not be correct since for example your laptop goes into hibernation and you connect back to the network it may show online but you are offline, and it will show offline before coming back up as you are disconnected and connect back ... I'm not sure that keeping the keep alive going will help.

>OTOH an unused idle backend should probably timeout.

I don't think we can do that, the moment the user has a terminal we cannot know if he wants to keep it alive or not. Even if it's not doing anything.

> To detect that sort of idleness a simple lack of user interaction would not be enough 

True but that was not the point, the point was to detect front-end idleness, not backed.

>Also, I don't like to see this information in the status bar as I don't think it is useful to the user. She probably knows best that she hasn't been active on that window. But also it doesn't matter, does it?.

Maybe true yes.

>I think this PR would be smaller and simpler if we left the 'idleness' out and simply focus on the offline/online status.

OK but I think we should bring back the code in another PR, still feels useful to me. And @kittaakos already has the code.

I actually don't see how this code is bad for the user or back-end operator ?

hexa00 commented:
@svenefftinge Moreover I don't know if you read the comments but @marcdumais-work and me found that useful, I think it would have been polite to discuss it with us before taking the decision and merging it while we were asleep.
